File: rdkit-sys/wrapper/src/bond.cc
#include "rust/cxx.h"
#include <GraphMol/GraphMol.h>

namespace RDKit {

using BondType   = Bond::BondType;
using BondStereo = Bond::BondStereo;
using BondDir    = Bond::BondDir;

unsigned int get_num_bonds(const std::shared_ptr<ROMol> &mol, bool only_heavy) { return mol->getNumBonds(only_heavy); }

Bond &get_bond_with_idx(std::shared_ptr<ROMol> &mol, unsigned int idx) { return *mol->getBondWithIdx(idx); }

int get_bond_idx_between_atoms(const std::shared_ptr<ROMol> &mol, unsigned int begin_idx, unsigned int end_idx) {
	const Bond *bond = mol->getBondBetweenAtoms(begin_idx, end_idx);
	if (bond == nullptr) { return -1; }
	return static_cast<int>(bond->getIdx());
}

BondType bond_get_bond_type(const Bond &bond) { return bond.getBondType(); }

double bond_get_bond_type_as_double(const Bond &bond) { return bond.getBondTypeAsDouble(); }

unsigned int bond_get_begin_atom_idx(const Bond &bond) { return bond.getBeginAtomIdx(); }

unsigned int bond_get_end_atom_idx(const Bond &bond) { return bond.getEndAtomIdx(); }

unsigned int bond_get_other_atom_idx(const Bond &bond, unsigned int this_idx) { return bond.getOtherAtomIdx(this_idx); }

bool bond_get_is_aromatic(const Bond &bond) { return bond.getIsAromatic(); }

bool bond_get_is_conjugated(const Bond &bond) { return bond.getIsConjugated(); }

BondStereo bond_get_stereo(const Bond &bond) { return bond.getStereo(); }

BondDir bond_get_bond_dir(const Bond &bond) { return bond.getBondDir(); }

unsigned int bond_get_idx(const Bond &bond) { return bond.getIdx(); }

} // namespace RDKit

File: rdkit-sys/src/bridge/bond.rs
#[cxx::bridge(namespace = "RDKit")]
pub mod ffi {
    #[repr(i32)]
    #[derive(Debug, PartialEq)]
    pub enum BondType {
        UNSPECIFIED,
        SINGLE,
        DOUBLE,
        TRIPLE,
        QUADRUPLE,
        QUINTUPLE,
        HEXTUPLE,
        ONEANDAHALF,
        TWOANDAHALF,
        THREEANDAHALF,
        FOURANDAHALF,
        FIVEANDAHALF,
        AROMATIC,
        IONIC,
        HYDROGEN,
        THREECENTER,
        DATIVEONE,
        DATIVE,
        DATIVEL,
        DATIVER,
        OTHER,
        ZERO,
    }

    #[repr(i32)]
    #[derive(Debug, PartialEq)]
    pub enum BondStereo {
        STEREONONE,
        STEREOANY,
        STEREOZ,
        STEREOE,
        STEREOCIS,
        STEREOTRANS,
    }

    #[repr(i32)]
    #[derive(Debug, PartialEq)]
    pub enum BondDir {
        NONE,
        BEGINWEDGE,
        BEGINDASH,
        ENDDOWNRIGHT,
        ENDUPRIGHT,
        EITHERDOUBLE,
        UNKNOWN,
    }

    unsafe extern "C++" {
        include!("wrapper/include/bond.h");

        pub type ROMol = crate::ro_mol_ffi::ROMol;
        pub type Bond;
        pub type BondType;
        pub type BondStereo;
        pub type BondDir;

        pub fn get_num_bonds(mol: &SharedPtr<ROMol>, only_heavy: bool) -> u32;
        pub fn get_bond_with_idx(mol: &mut SharedPtr<ROMol>, idx: u32) -> Pin<&mut Bond>;
        pub fn get_bond_idx_between_atoms(
            mol: &SharedPtr<ROMol>,
            begin_idx: u32,
            end_idx: u32,
        ) -> i32;

        pub fn bond_get_bond_type(bond: Pin<&Bond>) -> BondType;
        pub fn bond_get_bond_type_as_double(bond: Pin<&Bond>) -> f64;
        pub fn bond_get_begin_atom_idx(bond: Pin<&Bond>) -> u32;
        pub fn bond_get_end_atom_idx(bond: Pin<&Bond>) -> u32;
        pub fn bond_get_other_atom_idx(bond: Pin<&Bond>, this_idx: u32) -> u32;
        pub fn bond_get_is_aromatic(bond: Pin<&Bond>) -> bool;
        pub fn bond_get_is_conjugated(bond: Pin<&Bond>) -> bool;
        pub fn bond_get_stereo(bond: Pin<&Bond>) -> BondStereo;
        pub fn bond_get_bond_dir(bond: Pin<&Bond>) -> BondDir;
        pub fn bond_get_idx(bond: Pin<&Bond>) -> u32;
    }
}
